#1 - ما هي اللغات التي يوجد بها Garbage collection

In [ ]:
"""[java, c#, python, javascript, rupy, Go, PHP, Swift, R, Dart]"""

#2 - ما هم ال 13 مبدأ الخاص بكتابة clean code

In [ ]:
"""[1- MeaningFull Name, 2- Small Function, 3- Function Arguments, 4- Comments, 5- Code Structure,
6- Formating, 7- Error Handling, 8- Avoid Duplication, 9- Single Responsipilty Principle, 
10- Open Closed Principle, 11- Law of demeter, 12- Unit Testing, 13- Refactoring]"""

#3 - طريقة عمل do while في بايثون

In [1]:
while True:
    print("Hello Abdo.")
    
    condition = False 
    
    if not condition:
        break


Hello Abdo.


#4 - قارن بين for and while in python

In [ ]:
"""#- For - في بايثون، حلقات for غالبًا ما تكون أسرع من حلقات while عندما يكون عدد التكرارات معروفًا مسبقًا.
تعتبر أكثر كفاءة عند التكرار على كائنات قابلة للتكرار لأن بايثون تستخدم عملية مدمجة تسمى "iterator protocol" التي تكون محسنة للأداء.
عند استخدام حلقة for مع دالة range، يتم تنفيذ التكرار بكفاءة عالية."""
# -------------------------------
"""- While - يمكن أن تكون أبطأ من حلقات for في بعض الحالات بسبب الحاجة إلى التحقق من الشرط في كل تكرار.
تتطلب المزيد من الاهتمام لتجنب الحلقات غير المنتهية التي يمكن أن تؤدي إلى استهلاك غير مبرر للموارد."""

#5 - أشهر الأدوات اللي بتعمل Debugging في بايثون

In [ ]:
"""[1- PDB (Python Debugger): import pdb (pdb.set_trace()), 2- Jupyter Notebooks, 3- PyCharm Debugger, 
4- Visual Studio Code (VS Code), 5- Wing IDE, ]"""

#6- طريقة كتابة سطر بين سطور الفايل في بايثون

In [4]:
def insert_line_in_file(name_file, line_to_insert, line_number):
    with open(name_file, 'r') as file:
        lines = file.readlines()

    lines.insert(line_number, line_to_insert + '\n')

    with open(name_file, 'w') as file:
        file.writelines(lines)

name_file = 'taskai.txt'
line_to_insert = 'inserted New line.'
line_number = 1 

insert_line_in_file(name_file, line_to_insert, line_number)

"""Q31. Write a Python function square(num) that takes a number
as input and returns its square."""

In [5]:
def square_num(num):
    return num**2

num = int(input("Enter Number: "))

print(f"The Square Num {num} is: {square_num(num)}")

The Square Num 5 is: 25


"""Q32. Write a Python function sum_of_digits(number) that 
takes an integer as input and returns the sum of its digits."""

In [7]:
def sum_of_digits(num):
    num = abs(num)
    str_num = str(num)
    sum = 0
    for i in range(len(str_num)):
        sum += num % 10
        num = num // 10
    return sum

num = int(input("Enter Number: "))

print(f"The sum of digits num ({num}): {sum_of_digits(num)}")

The sum of digits num (123): 6


"""Q33. Write a Python function reverse_string(s) that takes a string
as input and returns the string in reverse order."""

In [8]:
def revers_string():
    string = input("Enter String: ")
    
    return string[::-1]

print(f"The Reverse String Is: {revers_string()}")

The Reverse String Is: demhA


"""Q34. Write a Python function count_vowels(s) that takes a string
as input and returns the count of vowels in the string."""

In [9]:
vowles_lst = ['a', 'e', 'i', 'o', 'u']

def coiunt_vowels(s):
    count = 0
    for char in s:
        if char.lower() in vowles_lst:
            count+=1

    return count


string = input("Enter a string: ")

print(f"Count Vowel in {string} = {coiunt_vowels(string)}")

Count Vowel in sky = 0


"""Q35. Write a Python function sort_list(lst) that takes
a list of numbers as input and returns a new list 
with the numbers sorted in ascending order."""

In [10]:
def sort_list(lst):
    new_lst = []
    lst.sort()
    new_lst = lst
    return new_lst

lst = []
while True:
    num = input("Enter Number: ")

    if num == 'done' or num == '':
        break
    
    num = int(num)

    lst.append(num)

print(f"The Old List Number: {lst}")
print("-"*25)
print(f"The New List Sorted Number: {sort_list(lst)}")

The Old List Number: [7, 9, 14, 2, 25]
-------------------------
The New List Sorted Number: [2, 7, 9, 14, 25]


"""Q36. Write a Python function word_count(sentence) that takes a sentence
as input and returns the number of words in the sentence."""

In [11]:
def word_count(sentence):
    word = sentence.split()

    return len(word)

sentence = input("Enter String: ")

print(f"The count of words is: {word_count(sentence)}")

The count of words is: 3


"""Q37. Write a Python function remove_elements(lst, n) that takes
a list and an index n as input and removes the element
at index n from the list."""

In [12]:
def remove_elemnts(lst, n):
    
    if 0 <= n < len(lst):
        lst.pop(n)
        return f"Index Number is remove {n} And\nThe New List: {lst}"
    else:
        return f"Not Found Index."


lst = []

while True:
    num = input("Enter Number To Append in List: ")

    if num == 'done' or num == '':
        break
    
    num = int(num)

    lst.append(num)


print(f"The List Before Remove Elemnt: {lst}")

num_remove = int(input("Enter The Number is Remove in List: "))

print(remove_elemnts(lst, num_remove))

The List Before Remove Elemnt: [1, 2, 3, 4]
Index Number is remove 1 And
The New List: [1, 3, 4]


"""Q38. Write a Python function duplicate_list(lst) that takes a list
as input and returns a new list that duplicates 
the elements of the original list."""

In [13]:
def dublicate_list(lst):
    new_lst = lst * 2

    return new_lst

lst = []
while True:
    
    num = input("Enter Number: ")

    if num == 'done' or num == '':
        break
    
    num = int(num)

    lst.append(num)

print(f"The Old List Number: {lst}")

print(f"The New List Dublicate Number: {dublicate_list(lst)}")

The Old List Number: [1, 2, 3]
The New List Dublicate Number: [1, 2, 3, 1, 2, 3]


"""Q39. Write a Python function is_divisible(num1, num2) 
that takes two numbers as input and returns 
True if the first number is divisible by the second number, 
and False otherwise."""

In [14]:
def is_divisible():
    num1 = int(input("Enter Number 1: "))
    num2 = int(input("Enter Number 2: "))

    if num1 % num2 == 0:
        return True
    else:
        return False

print(is_divisible())

True


"""Q40. Write a Python function power(base, exponent) that takes 
a base and an exponent as input and returns 
the result of the base raised to the exponent."""

In [15]:
def power():
    base = int(input("Enter The Base: "))
    exponent = int(input("Enter The Exponent: "))
    return pow(base, exponent)

print(power())

4


Q41. Write a program which will find all such numbers which are divisible by 7 but are not divisible by 5, between 2000 and 3000 (both included). The numbers obtained should be printed in a comma-separated sequence on a single line.

In [5]:
result = []

for num in range(2000, 3001):
    if (num % 7 == 0) and (num % 5 != 0):
        result.append(str(num))

print(",".join(result))


2002,2009,2016,2023,2037,2044,2051,2058,2072,2079,2086,2093,2107,2114,2121,2128,2142,2149,2156,2163,2177,2184,2191,2198,2212,2219,2226,2233,2247,2254,2261,2268,2282,2289,2296,2303,2317,2324,2331,2338,2352,2359,2366,2373,2387,2394,2401,2408,2422,2429,2436,2443,2457,2464,2471,2478,2492,2499,2506,2513,2527,2534,2541,2548,2562,2569,2576,2583,2597,2604,2611,2618,2632,2639,2646,2653,2667,2674,2681,2688,2702,2709,2716,2723,2737,2744,2751,2758,2772,2779,2786,2793,2807,2814,2821,2828,2842,2849,2856,2863,2877,2884,2891,2898,2912,2919,2926,2933,2947,2954,2961,2968,2982,2989,2996


### 42. Write a program which can compute the factorial of a given numbers. The results should be printed in a comma-separated sequence on a single line.
    5 8
    Then, the output should be:
    120,40320

In [9]:
def factorial(n):
    if n == 0:
        return 1
    else:
        return n * (factorial(n - 1))
    
numbers = input("Enter Numbers: ")

num_lst = numbers.split(",")

factorials = [str(factorial(int(num))) for num in num_lst]

print(",".join(factorials))

120,40320


### 43. With a given integral number n, write a program to generate a dictionary that contains (i, i*i) such that is an integral number between 1 and n (both included). and then the program should print the dictionary.
    Suppose the following input is supplied to the program:
    8
    Then, the output should be:
    {1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64}

In [10]:
num = int(input("Enter Num: "))

num_dict = {i: i*i for i in range(1, num+1)}

print(num_dict)

{1: 1, 2: 4, 3: 9, 4: 16, 5: 25, 6: 36, 7: 49, 8: 64}


### 44. Write a program which accepts a sequence of comma-separated numbers from console and generate a list and a tuple which contains every number. Suppose the following input is supplied to the program:
    34,67,55,33,12,98
    Then, the output should be:
    ['34', '67', '55', '33', '12', '98']
    ('34', '67', '55', '33', '12', '98')

In [12]:
numbers = input("Enter Numebrs: ")

num_lst = numbers.split(",")

t_num = tuple(num_lst)

print(num_lst)
print(t_num)

['34', '67', '55', '33', '12', '98']
('34', '67', '55', '33', '12', '98')


### 45. Write a program that calculates and prints the value according to the given formula:
    Q = Square root of [(2 * C * D) / H], 
        C is 50. H is 30.
        D is the variable whose values should be input to your program in a comma-separated sequence.

- Example
    - Let us assume the following comma separated input sequence is given to the program:
        100,150,180
    - The output of the program should be:
        18,22,24

In [16]:
import math

c = 50

h = 30

num_input = input("Enter Number")

D_num = num_input.split(",")


new_lst = []


for D in D_num:
    D = int(D)
    Q = math.sqrt((2 * c * D) / h)
    new_lst.append(str(int(Q)))

print(",".join(new_lst))

18,22,24


### 46. Write a program which takes 2 digits, X,Y as input and generates a 2-dimensional array. The element value in the i-th row and j-th column of the array should be i*j.
    Example
        Suppose the following inputs are given to the program:
        3,5
        Then, the output of the program should be:
        [[0, 0, 0, 0, 0], [0, 1, 2, 3, 4], [0, 2, 4, 6, 8]]

In [17]:
input_num = input("Enter X,Y: ")

X, Y = map(int, input_num.split(","))

array = [[i*j for j in range(Y)] for i in range(X)]

for rwo in array:
    print(rwo)

[0, 0, 0, 0, 0]
[0, 1, 2, 3, 4]
[0, 2, 4, 6, 8]


### 47. Write a program that accepts a sequence of whitespace separated words as input and prints the words after removing all duplicate words and sorting them alphanumerically. Suppose the following input is supplied to the program:
    hello world and practice makes perfect and hello world again
    Then, the output should be:
    again and hello makes perfect practice world

In [22]:
str_input = input("Enter The string: ")

lst_str = str_input.split(" ")

set_str = set(lst_str)

lst_sorted = sorted(set_str)
print(" ".join(lst_sorted))

again and hello makes perfect practice world


### 48.Write a program which accepts a sequence of comma separated 4 digit binary numbers as its input and then check whether they are divisible by 5 or not. The numbers that are divisible by 5 are to be printed in a comma separated sequence.

    Example:
    0100,0011,1010,1001
    Then the output should be:
    1010

In [4]:
input_num = input("Enter num: ")

lst_split = input_num.split(",")

lst_num = []

for num in lst_split:
    if int(num, 2) % 5 == 0:
        lst_num.append(num)


print(",".join(lst_num))


1010


### 49. Write a program, which will find all such numbers between 1000 and 3000 (both included) such that each digit of the number is an even number. The numbers obtained should be printed in a comma-separated sequence on a single line.

In [9]:
num_lst = []

for num in range(1000, 3001):
    if all(int(digit) % 2 == 0 for digit in str(num)):
        num_lst.append(str(num))

print(",".join(num_lst))

2000,2002,2004,2006,2008,2020,2022,2024,2026,2028,2040,2042,2044,2046,2048,2060,2062,2064,2066,2068,2080,2082,2084,2086,2088,2200,2202,2204,2206,2208,2220,2222,2224,2226,2228,2240,2242,2244,2246,2248,2260,2262,2264,2266,2268,2280,2282,2284,2286,2288,2400,2402,2404,2406,2408,2420,2422,2424,2426,2428,2440,2442,2444,2446,2448,2460,2462,2464,2466,2468,2480,2482,2484,2486,2488,2600,2602,2604,2606,2608,2620,2622,2624,2626,2628,2640,2642,2644,2646,2648,2660,2662,2664,2666,2668,2680,2682,2684,2686,2688,2800,2802,2804,2806,2808,2820,2822,2824,2826,2828,2840,2842,2844,2846,2848,2860,2862,2864,2866,2868,2880,2882,2884,2886,2888


### 50. Write a program that accepts a sentence and calculate the number of letters and digits. Suppose the following input is supplied to the program:

    "hello world 123"
    Then, the output should be:
        LETTERS 10
        DIGITS 3

In [3]:
sentence = input("Enter The Word: ")

letter_count = 0
letter_digit = 0

for char in sentence:
    if char.isalpha():
        letter_count += 1
    elif char.isdigit():
        letter_digit += 1

print("LETTERS:",letter_count)
print("DIGITS:",letter_digit)



LETTERS: 10
DIGITS: 3


### 51. Write a program that accepts a sentence and calculate the number of upper case letters and lower case letters. Suppose the following input is supplied to the program:

    Hello World
    Then, the output should be:
        UPPER CASE 2
        LOWER CASE 8

In [4]:
sentence = input("Enter The Word: ")

count_upper = 0
count_lower = 0

for char in sentence:
    if char.isupper():
        count_upper += 1
    elif char.islower():
        count_lower += 1

print("UPPER CASE:",count_upper)
print("LOWER CASE:",count_lower)

UPPER CASE: 2
LOWER CASE: 8


### 52. Use a list comprehension to square each odd number in a list. The list is input by a sequence of comma-separated numbers. Suppose the following input is supplied to the program:
    1,2,3,4,5,6,7,8,9
    Then, the output should be:
    1,3,5,7,9

In [8]:
inp_str = input("input by a sequence of comma-separated numbers: ")

numbers = list(map(int, inp_str.split(",")))

num_odd = [num for num in numbers if num % 2 != 0]

print(*num_odd, sep = ",")

1,3,5,7,9


### 53. You are required to write a program to sort the (name, age, height) tuples by ascending order where name is string, age and height are numbers. The tuples are input by console. The sort criteria is:
    1: Sort based on name;
    2: Then sort based on age;
    3: Then sort by score.
    The priority is that name > age > score.

    If the following tuples are given as input to the program:
        Tom,19,80
        John,20,90
        Jony,17,91
        Jony,17,93
        Json,21,85
    Then, the output of the program should be:
    [('John', '20', '90'), ('Jony', '17', '91'), ('Jony', '17', '93'), ('Json', '21', '85'), ('Tom', '19', '80')]

In [9]:
tuble_list = []

print("Enter the tuples (name, age, height), one per line (press Enter to stop):")

while True:
    
		try:
			user_input = input()

			if not user_input:
				break

			name,age,height = user_input.split(",")
			tuble_list.append((name, age, height))
		except EOFError:
			break

sorted_list = sorted(tuble_list, key = lambda x: (x[0], int(x[1]), int(x[2])))

print(sorted_list)

Enter the tuples (name, age, height), one per line (press Enter to stop):
[('John', '20', '90'), ('Jony', '17', '91'), ('Jony', '17', '93'), ('Json', '21', '85'), ('Tom', '19', '18')]


### 54. Define a function that can accept an integer number as input and print the "It is an even number" if the number is even, otherwise print "It is an odd number".

In [12]:
def num_even_or_odd(num):
    if num % 2 == 0:
        print(f"The Num {num} is Even.")
    else:
        print(f"The Num {num} is Odd.")

num = int(input("Enter The num: "))

num_even_or_odd(num)

The Num 7 is Odd.


### 55.The Fibonacci Sequence is computed based on the following formula:
    f(n)=0 if n=0
    f(n)=1 if n=1
    f(n)=f(n-1)+f(n-2) if n>1

- Please write a program using list comprehension to print the Fibonacci Sequence in comma separated form with a given n input by console.

- Example:

    If the following n is given as input to the program:
    7
    Then, the output of the program should be:
    0,1,1,2,3,5,8,13

In [21]:
def fibonacci_sequence(n):
    if n == 0:
        return []
    elif n == 1:
        return [0]
    else:
        lst = [0, 1]
        [lst.append(lst[-1] + lst[-2]) for _ in range(2, n)]
        return lst

num = int(input("Enter number: "))

fipo = fibonacci_sequence(num)

print(",".join(map(str, fipo)))

0,1,1,2,3,5,8


### Q56. Define a Python class named 'Person' with attributes 'name' and 'age'.

In [1]:
class Person:
    def __init__(self, name, age):
        self.name = name
        self.age = age
    

P1 = Person("Abdo", 22)

print(f"Name: {P1.name}, Age: {P1.age}")

Name: Abdo, Age: 22


### Q57. Add a constructor to the 'Person' class that initializes the 'name' and 'age' attributes.

In [2]:
class Person:
    def __init__(self, name, age):
        self.name = name
        self.age = age
    

P1 = Person("Abdo", 22)
P2 = Person("Ahmed", 24)

print(f"Name: {P1.name}, Age: {P1.age}")
print(f"Name: {P2.name}, Age: {P2.age}")

Name: Abdo, Age: 22
Name: Ahmed, Age: 24


### Q58. Create an instance method in the 'Person' class that prints the person's information.

In [1]:
class Person:
    def __init__(self, name, age):
        self.name = name
        self.age = age
    
    def print_info(self):
        print(f"Name: {self.name}, Age: {self.age}")

P1 = Person("Abdo", 22)
P2 = Person("Ahmed", 24)

P1.print_info()
P2.print_info()


Name: Abdo, Age: 22
Name: Ahmed, Age: 24


### Q59. Create an object of the 'Person' class and call the instance method to display its information.

In [2]:
class Person:
    def __init__(self, name, age):
        self.name = name
        self.age = age
    
    def print_info(self):
        print(f"Name: {self.name}, Age: {self.age}")

Person1 = Person("Abdo", 22)

Person1.print_info()

Name: Abdo, Age: 22


### Q60. Create a subclass 'Student' that inherits from the 'Person' class, and add a new attribute 'student_id'.

In [5]:
class Person:
    def __init__(self, name, age):
        self.name = name
        self.age = age
    
    def print_info(self):
        print(f"Name: {self.name}, Age: {self.age}")

class Student(Person):
    def __init__(self,name, age, student_id):
        super().__init__(name, age)
        self.student_id = student_id
    def print_info(self):
        super().print_info()
        print(f"student ID: {self.student_id}")


student1 = Student("Abdo", 22, 42022311)

student1.print_info()

Name: Abdo, Age: 22
student ID: 42022311
